python/pyiceberg/expressions/visitors.py

```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
+
+        return visit(self.expr, self)
+
+    def _contains_nulls_only(self, idx: int) -> bool:
+        return idx in self.value_counts and idx in self.null_counts and self.value_counts[idx] - self.null_counts[idx] == 0
+
+    def _contains_nans_only(self, idx: int) -> bool:
+        return idx in self.nan_counts and idx in self.value_counts and self.nan_counts[idx] == self.value_counts[idx]
+
+    def _is_nan(self, val: Any) -> bool:
+        return math.isnan(val)
+
+    def visit_true(self) -> bool:
+        # all rows match
+        return ROWS_MIGHT_MATCH
+
+    def visit_false(self) -> bool:
+        # all rows fail
+        return ROWS_CANNOT_MATCH
+
+    def visit_not(self, child_result: bool) -> bool:
+        raise ValueError(f"NOT should be rewritten: {child_result}")
+
+    def visit_and(self, left_result: bool, right_result: bool) -> bool:
+        return left_result and right_result
+
+    def visit_or(self, left_result: bool, right_result: bool) -> bool:
+        return left_result or right_result
+
+    def visit_is_null(self, term: BoundTerm[L]) -> bool:
+        idx = term.ref().field.field_id
+
+        if self.null_counts is not None and idx in self.null_counts and self.null_counts[idx] == 0:
+            return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_not_null(self, term: BoundTerm[L]) -> bool:
+        # no need to check whether the field is required because binding evaluates that case
+        # if the column has no non-null values, the expression cannot match
+        idx = term.ref().field.field_id
+
+        if self._contains_nulls_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_is_nan(self, term: BoundTerm[L]) -> bool:
+        idx = term.ref().field.field_id
+
+        if idx in self.nan_counts and self.nan_counts[idx] == 0:
+            return ROWS_CANNOT_MATCH
+
+        # when there's no nanCounts information, but we already know the column only contains null,
+        # it's guaranteed that there's no NaN value
+        if self._contains_nulls_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_not_nan(self, term: BoundTerm[L]) -> bool:
+        idx = term.ref().field.field_id
+
+        if self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_less_than(self, term: BoundTerm[L], literal: Literal[L]) -> bool:
+        field = term.ref().field
+        idx = field.field_id
+
+        if self._contains_nulls_only(idx) or self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        if not isinstance(field.field_type, PrimitiveType):
+            raise ValueError(f"Expected PrimitiveType: {field.field_type}")
+
+        if idx in self.lower_bounds:
+            lower = from_bytes(field.field_type, self.lower_bounds[idx])
+
+            if self._is_nan(lower):
```

Comment:
This could potentially pass `None` and non-float values. We should make sure that `_is_nan` handles those correctly.

Reply:
Good call, I've updated this to:
```python
    def _is_nan(self, val: Any) -> bool:
        try:
            return math.isnan(val)
        except TypeError:
            # In the case of None or other non-numeric types
            return False
```